File: app/src/main/java/workshop1024/com/xproject/controller/fragment/HomePageFragment.java
package workshop1024.com.xproject.controller.fragment;

import android.os.Bundle;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentStatePagerAdapter;
import android.support.v4.view.ViewPager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import workshop1024.com.xproject.R;

/**
 * 抽屉导航Home Fragment，包含ViewPager来显示Stories和Topies子PageFragment
 */
public class HomePageFragment extends TopFragment {

    public HomePageFragment() {
        // Required empty public constructor
    }

    public static HomePageFragment newInstance() {
        HomePageFragment homePageFragment = new HomePageFragment();
        homePageFragment.setNavigationItemId(R.id.leftnavigator_menu_home);
        return homePageFragment;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.homepage_fragment, container, false);

        TabLayout tabLayout = view.findViewById(R.id.homepage_tablayout_tabs);
        ViewPager viewPager = view.findViewById(R.id.homepage_viewpager_fragments);

        //不使用getChildFragmentManager，从StoryFragment返回，PageFragment不显示
        HomeFragmentPagerAdapter homeFragmentPagerAdapter = new HomeFragmentPagerAdapter(getChildFragmentManager());
        viewPager.setAdapter(homeFragmentPagerAdapter);
        tabLayout.setupWithViewPager(viewPager, true);

        return view;
    }

    /**
     * HomeFragment的ViewPager适配器
     */
    public class HomeFragmentPagerAdapter extends FragmentStatePagerAdapter {
        private String[] mTabTitles;

        public HomeFragmentPagerAdapter(FragmentManager fragmentManager) {
            super(fragmentManager);
            mTabTitles = getResources().getStringArray(R.array.homepage_tabs_strings);
        }

        @Override
        public Fragment getItem(int position) {
            return PageAdapterFragment.newInstance();
        }

        @Override
        public int getCount() {
            return mTabTitles.length;
        }

        @Override
        public CharSequence getPageTitle(int position) {
            return mTabTitles[position];
        }
    }
}

File: home/src/main/java/workshop1024/com/xproject/home/controller/fragment/HomePageFragment.kt
package workshop1024.com.xproject.home.controller.fragment

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentStatePagerAdapter
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import workshop1024.com.xproject.base.controller.fragment.TopFragment
import workshop1024.com.xproject.base.controller.fragment.XFragment

import workshop1024.com.xproject.home.R
import workshop1024.com.xproject.base.controller.event.HomePageAsReadEvent
import workshop1024.com.xproject.base.controller.event.HomePageRefreshEvent
import workshop1024.com.xproject.home.databinding.HomepageFragmentBinding

/**
 * 抽屉导航Home Fragment，包含ViewPager来显示Stories和Topies子PageFragment
 */
class HomePageFragment : XFragment(), TopFragment {
    private var mTabTitles: Array<String>? = null

    private var mHomeFragmentPagerAdapter: HomeFragmentPagerAdapter? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val homepageFragmentBinding = DataBindingUtil.inflate<HomepageFragmentBinding>(inflater, R.layout.homepage_fragment,
                container, false)

        //设置ViewPager允许有所有的存在屏幕外的Fragment不会被销毁
        mTabTitles = resources.getStringArray(R.array.homepage_tabs_strings)
        homepageFragmentBinding.homepageViewpagerFragments.offscreenPageLimit = mTabTitles!!.size - 1

        //不使用getChildFragmentManager，从StoryFragment返回，PageFragment不显示
        mHomeFragmentPagerAdapter = HomeFragmentPagerAdapter(childFragmentManager)
        homepageFragmentBinding.homepageViewpagerFragments.adapter = mHomeFragmentPagerAdapter
        homepageFragmentBinding.homepageTablayoutTabs.setupWithViewPager(homepageFragmentBinding.homepageViewpagerFragments,
                true)

        return homepageFragmentBinding.root
    }

    override fun onStart() {
        super.onStart()
        EventBus.getDefault().register(this)
    }

    override fun onStop() {
        EventBus.getDefault().unregister(this)
        super.onStop()
    }

    @Subscribe
    fun onRefreshEvent(event: HomePageRefreshEvent) {
        mHomeFragmentPagerAdapter?.mCurrentSubFragmet?.onRefresh()
    }

    @Subscribe
    fun onMarkAsReadEvent(event: HomePageAsReadEvent) {
        mHomeFragmentPagerAdapter?.mCurrentSubFragmet?.markAsRead()
    }

    inner class HomeFragmentPagerAdapter(fragmentManager: FragmentManager) : FragmentStatePagerAdapter(fragmentManager) {
        var mCurrentSubFragmet: HomeSubFragment? = null
            private set

        override fun getItem(position: Int): Fragment {
            var fragment: Fragment? = null

            when (position) {
                0 -> //FIXME 是不是有必要每次都Nes一个Fragment对象
                    fragment = SubscribeFragment.newInstance()
                1 -> fragment = TagFragment.newInstance()
                2 -> fragment = FilterFragment.newInstance()
            }

            return fragment!!
        }

        override fun getCount(): Int {
            return mTabTitles!!.size
        }

        override fun getPageTitle(position: Int): CharSequence {
            return mTabTitles!![position]
        }

        override fun setPrimaryItem(container: ViewGroup, position: Int, `object`: Any) {
            mCurrentSubFragmet = `object` as HomeSubFragment
            super.setPrimaryItem(container, position, `object`)
        }
    }

    companion object {
        fun newInstance(navigationItemId: Int): HomePageFragment {
            val homePageFragment = HomePageFragment()
            homePageFragment.mNavigationItemId = navigationItemId
            return homePageFragment
        }
    }
}